File: data_structures/trees/heap/python/heap.py
from typing import List, Generic, TypeVar, Callable, Final, Union
import math

T = TypeVar("T")


def default_comparator(a: int, b: int) -> int:
    return a - b


def get_parent_index(i: int) -> int:
    return math.ceil(i / 2 - 1)


def get_left_index(i: int) -> int:
    return 2 * i + 1


def get_right_index(i: int) -> int:
    return 2 * i + 2


class Heap(Generic[T]):
    comparator: Callable[[int, int], int]
    heap: List[T]

    def __init__(self, comparator: Callable[[int, int], int] = default_comparator):
        self.heap = []

        def construct_comparator(i1: int, i2: int) -> int:
            value = comparator(self.heap[i1], self.heap[i2])

            if not isinstance(value, int):
                raise ValueError("Comparator should evaluate to a number")

            return value

        self.comparator = construct_comparator

    def add(self, value: T) -> None:
        self.heap.append(value)
        self.bubble_up()

    def bubble_up(self) -> None:
        index = self.size - 1

        while get_parent_index(index) >= 0 and self.comparator(get_parent_index(index), index) > 0:
            parent_index = get_parent_index(index)
            self.swap(parent_index, index)
            index = parent_index

    def bubble_down(self, index: int = 0) -> None:
        curr = index

        while get_left_index(curr) < self.size and self.comparator(curr, self.get_child(curr)) > 0:
            next_element = self.get_child(curr)
            self.swap(curr, next_element)
            curr = next_element

    def remove(self, index: int = 0) -> Union[T, None]:
        if not self.size:
            return None

        self.swap(index, self.size - 1)
        value: Final = self.heap.pop()
        self.bubble_down(index)

        return value

    def get_child(self, i: int) -> int:
        if get_right_index(i) < self.size and self.comparator(get_left_index(i), get_right_index(i)) > 0:
            return get_right_index(i)

        return get_left_index(i)

    def peek(self) -> T:
        return self.heap[0]

    def swap(self, i1: int, i2: int) -> None:
        self.heap[i1], self.heap[i2] = self.heap[i2], self.heap[i1]

    @property
    def size(self) -> int:
        return len(self.heap)

File: data_structures/trees/heap/typescript/Heap.ts
export type ComparatorFunction = (a, b) => number;

/**
 * Returns parents index
 * @param index
 */
const getParentIndex = (index: number): number => Math.ceil(index / 2 - 1);
const getLeftIndex = (i: number): number => 2 * i + 1;
const getRightIndex = (i: number): number => 2 * i + 2;

export default class Heap<T> {
    comparator: ComparatorFunction;
    heap: T[] = [];

    constructor(comparator: ComparatorFunction = (a: number, b: number) => a - b) {
        this.heap = [];
        this.comparator = (i1: number, i2: number) => {
            const value = comparator(this.heap[i1], this.heap[i2]);
            if (Number.isNaN(value)) {
                throw new Error(`Comparator should evaluate to a number`);
            }
            return value;
        };
    }

    /**
     * Insert element
     * @runtime O(log n)
     */
    add(value: T): void {
        this.heap.length = this.size + 1;
        this.heap[this.size - 1] = value;
        this.bubbleUp();
    }

    /**
     * Retrieves, but does not remove, the head of this heap
     * @runtime O(1)
     */
    peek(): T {
        return this.heap[0];
    }

    /**
     * Retrieves and removes the head of this heap, or returns null if this heap is empty
     * @runtime O(log n)
     */
    remove(index = 0): T {
        if (!this.size) {
            return null;
        }

        const lastIndex = this.size - 1;
        this.swap(index, lastIndex); // swap with last
        const value = this.heap[lastIndex];
        this.heap.length = this.size - 1; // remove last element
        this.bubbleDown(index);

        return value;
    }

    /**
     * Returns the number of elements in this collection.
     * @runtime O(1)
     */
    get size(): number {
        return this.heap.length;
    }

    /**
     * Move new element upwards on the heap, if it's out of order
     * @runtime O(log n)
     */
    private bubbleUp(): void {
        let index = this.size - 1;

        while (getParentIndex(index) >= 0 && this.comparator(getParentIndex(index), index) > 0) {
            const parentIndex = getParentIndex(index);
            this.swap(parentIndex, index);
            index = parentIndex;
        }
    }

    /**
     * After removal, moves element downwards on the heap, if it's out of order
     * @runtime O(log n)
     */
    private bubbleDown(index = 0): void {
        let curr = index;

        while (getLeftIndex(curr) < this.size && this.comparator(curr, this.getChild(curr)) > 0) {
            const next = this.getChild(curr);
            this.swap(curr, next);
            curr = next;
        }
    }

    getChild(i: number): number {
        if (getRightIndex(i) < this.size && this.comparator(getLeftIndex(i), getRightIndex(i)) > 0) {
            return getRightIndex(i);
        }

        return getLeftIndex(i);
    }

    /**
     * Swap elements on the heap
     * @runtime O(1)
     * @param {number} i1 index 1
     * @param {number} i2 index 2
     */
    swap(i1: number, i2: number): void {
        [this.heap[i1], this.heap[i2]] = [this.heap[i2], this.heap[i1]];
    }
}
